### PR TITLE
[12.0][ADD] stock_card_report

### PR DESCRIPTION
**Pull request by ps-tubtim**

### CODE REVIEW

**Comment by rousseldenis** [comment]
@ps-tubtim @kittiu I like this kind of report.

For a customer of mine, this would display lot traceability.

Should we maybe modify it to help overriding queries. I will try to make a PR on your repo.

**Comment by rousseldenis** [comment]
> Should we maybe modify it to help overriding queries. I will try to make a PR on your repo.

As this module could be a base for reporting

**Comment by kittiu** [comment]
@rousseldenis sure.
cc @ps-tubtim

**Comment by rousseldenis** [comment]
@kittiu Is it the same usage of #42 ?

**Comment by kittiu** [comment]
> @kittiu Is it the same usage of #42 ?

From what I seen from the report columns, it looks different.

**Comment by rousseldenis** [comment]
The track should also lead to this :
https://github.com/OCA/mis-builder

as it allows to define reporting on every Odoo object.

**Comment by kittiu** [comment]
> The track should also lead to this :
> https://github.com/OCA/mis-builder
> 
> as it allows to define reporting on every Odoo object.

Confused by your point. :P what are you saying?

**Comment by kittiu** [comment]
@rousseldenis so, can you explain your point more? :)

**Comment by rousseldenis** [comment]
@kittiu I was in a questionning about reporting in general and to find the best method to do it. Maybe, that solution (mis-builder) is a little bit too arsh. I'll try to review this ASAP.

**Review by kittiu — APPROVED** [review]

**Comment by ps-tubtim** [comment]
@newtratip @Saran440 Could you review this PR?

File: stock_card_report/reports/stock_card_report.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                and move.state = 'done' and move.product_id in %s
+                and CAST(move.date AS date) <= %s
+            ORDER BY move.date, move.reference
+        """, (
+            tuple(locations.ids), tuple(locations.ids), self.date_from,
+            tuple(locations.ids), tuple(locations.ids),
+            tuple(self.product_ids.ids+[0]), self.date_to))
+        stock_card_results = self._cr.dictfetchall()
+        ReportLine = self.env['stock.card.view']
+        for line in stock_card_results:
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report.py]:
```suggestion
        self.results = [ReportLine.new(line).id for line in stock_card_results]
```
performance faster than a little bit

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                    'format': self.format_tcell_amount_right,
+                },
+                'width': 20,
+            },
+        }
+
+        ws_params = {
+            'ws_name': product.name,
+            'generate_ws_method': '_stock_card_report',
+            'title': 'Stock Card Report - '+product.name,
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report_xlsx.py]:
```suggestion
            'title': 'Stock Card Report - %s' % product.name,
```

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        ws.set_header(self.xls_headers['standard'])
+        ws.set_footer(self.xls_footers['standard'])
+
+        self._set_column_width(ws, ws_params)
+
+        row_pos = 0
+        row_pos = self._write_ws_title(ws, row_pos, ws_params, True)
+
+        for o in objects:
+            ws.write_row(
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report_xlsx.py]: please write filter by used function _write_line
https://github.com/OCA/reporting-engine/pull/364

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        ws.fit_to_pages(1, 0)
+        ws.set_header(self.xls_headers['standard'])
+        ws.set_footer(self.xls_footers['standard'])
+
+        self._set_column_width(ws, ws_params)
+
+        row_pos = 0
+        row_pos = self._write_ws_title(ws, row_pos, ws_params, True)
+
+        for o in objects:
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report_xlsx.py]:
```suggestion
        for o in objects.results:
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        self.date_from = self.date_range_id.date_start
+        self.date_to = self.date_range_id.date_end
+
+    @api.multi
+    def button_export_html(self):
+        self.ensure_one()
+        action = self.env.ref(
+            'stock_card_report.action_report_stock_card_report_html')
+        vals = action.read()[0]
+        context1 = vals.get('context', {})
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
        context = vals.get('context', {})
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        if isinstance(context1, pycompat.string_types):
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
        if isinstance(context, pycompat.string_types):
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            context1 = safe_eval(context1)
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
            context = safe_eval(context)
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        model = self.env['report.stock.card.report']
+        report = model.create(self._prepare_stock_card_report())
+        context1['active_id'] = report.id
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
        context['active_id'] = report.id
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        context1['active_ids'] = report.ids
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
        context['active_ids'] = report.ids
```

File: stock_card_report/wizard/stock_card_report_wizard.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        vals['context'] = context1
```
Comment (newtratip) [inline stock_card_report/wizard/stock_card_report_wizard.py]:
```suggestion
        vals['context'] = context
```

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            ws.write_row(row_pos+1, 2, [o.location_id.name or ''],
+                         self.format_tcell_center)
+
+            row_pos += 3
+            row_pos = self._write_line(
+                ws, row_pos, ws_params, col_specs_section='header',
+                default_format=self.format_theader_blue_center)
+            ws.freeze_panes(row_pos, 0)
+
+            balance = o._get_initial(o.results.filtered(
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report_xlsx.py]:
```suggestion
            balance = self._get_initial(o.filtered(
```

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                lambda l: l.product_id == product and l.is_initial))
+            ws.write_row(row_pos, 0, ['', 'Initial', '', ''],
```
Comment (Saran440) [inline stock_card_report/reports/stock_card_report_xlsx.py]: used _write_line function

**Review by Saran440 — COMMENTED** [review]
Code Reviewed.

File: stock_card_report/reports/stock_card_report.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    @api.multi
+    def _compute_results(self):
+        self.ensure_one()
+        change = False
+        if not self.date_from:
+            change = True
+            self.date_from = '0001-01-01'
+        if not self.date_to:
+            self.date_to = fields.Date.context_today(self)
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report.py]: **change** variable for update only self.date_from is null in last line of this function ? I think you should to declare variable date_from for self.date_from and date_to for self.date_to to better and short code.

**This is example:**
date_from = self.date_from or '0001-01-01'
date_to = self.date_to or fields.Date.context_today(self)

File: stock_card_report/reports/stock_card_report.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+                case when move.date < %s then True else False end as is_initial
+            FROM stock_move move
+            WHERE (move.location_id in %s or move.location_dest_id in %s)
+                and move.state = 'done' and move.product_id in %s
+                and CAST(move.date AS date) <= %s
+            ORDER BY move.date, move.reference
+        """, (
+            tuple(locations.ids), tuple(locations.ids), self.date_from,
+            tuple(locations.ids), tuple(locations.ids),
+            tuple(self.product_ids.ids+[0]), self.date_to))
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report.py]: Why must +[0] in self.product_ids.ids
```suggestion
            tuple(self.product_ids.ids), date_to))
```

File: stock_card_report/reports/stock_card_report.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+                case when move.location_id in %s
+                    then move.product_qty end as product_out,
+                case when move.date < %s then True else False end as is_initial
+            FROM stock_move move
+            WHERE (move.location_id in %s or move.location_dest_id in %s)
+                and move.state = 'done' and move.product_id in %s
+                and CAST(move.date AS date) <= %s
+            ORDER BY move.date, move.reference
+        """, (
+            tuple(locations.ids), tuple(locations.ids), self.date_from,
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report.py]:
```suggestion
            tuple(locations.ids), tuple(locations.ids), date_from,
```

File: stock_card_report/reports/stock_card_report.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+            tuple(locations.ids), tuple(locations.ids),
+            tuple(self.product_ids.ids+[0]), self.date_to))
+        stock_card_results = self._cr.dictfetchall()
+        ReportLine = self.env['stock.card.view']
+        for line in stock_card_results:
+            self.results += ReportLine.new(line)
+        if change:
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report.py]:
```suggestion
        
```

File: stock_card_report/reports/stock_card_report.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+            self.date_from = not change
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report.py]:
```suggestion
            
```

File: stock_card_report/reports/stock_card_report_xlsx.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
+# Copyright 2019 Ecosoft Co., Ltd. (http://ecosoft.co.th)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from odoo import models
+
+_logger = logging.getLogger(__name__)
+
+
+class ReportStockCardReportXlsx(models.TransientModel):
```
Comment (newtratip) [inline stock_card_report/reports/stock_card_report_xlsx.py]: models.AbstractModel ?

models.TransientModel will created table in the database.

Selection_073

File: stock_card_report/reports/stock_card_report_xlsx.py (hunk 3 above)
Comment (newtratip) [reply stock_card_report/reports/stock_card_report_xlsx.py]: @Saran440 What is the difference between write_row and _write_line ?

File: stock_card_report/reports/stock_card_report.py (hunk 1 above)
Comment (ps-tubtim) [reply stock_card_report/reports/stock_card_report.py]: Changed.

File: stock_card_report/reports/stock_card_report_xlsx.py (hunk 2 above)
Comment (ps-tubtim) [reply stock_card_report/reports/stock_card_report_xlsx.py]: I think I should to use this code `'title': 'Stock Card Report - {}'.format(product.name),`.

File: stock_card_report/wizard/stock_card_report_wizard.py (hunk 5 above)
Comment (ps-tubtim) [reply stock_card_report/wizard/stock_card_report_wizard.py]: Changed.

File: stock_card_report/reports/stock_card_report.py (hunk 13 above)
Comment (ps-tubtim) [reply stock_card_report/reports/stock_card_report.py]: Changed

File: stock_card_report/reports/stock_card_report.py (hunk 14 above)
Comment (ps-tubtim) [reply stock_card_report/reports/stock_card_report.py]: Edited

File: stock_card_report/reports/stock_card_report_xlsx.py (hunk 18 above)
Comment (ps-tubtim) [reply stock_card_report/reports/stock_card_report_xlsx.py]: I changed `models.TransientModel` to `models.AbstractModel`.

**Comment by ps-tubtim** [comment]
@Saran440 @newtratip Thank you for review.

**Comment by newtratip** [comment]
Hello @ps-tubtim 
I test by select multiple product as picture below 
Selection_074
and click view, the result will show
Selection_075

Customizable Desk **is not show Date To**
Corner Desk Right Sit **is show Date To**
Large Cabinet **is show Date To**

Please check that Date To is not same with all product.

**Comment by ps-tubtim** [comment]
@newtratip I edited this comment https://github.com/OCA/stock-logistics-reporting/pull/77#issuecomment-601020382. Please test again.

**Review by Saran440 — APPROVED** [review]
Code Reviewed. LGTM :+1:

**Review by newtratip — APPROVED** [review]
It seems OK, Waiting another review :)

**Pull request merged**